File: src/main/java/me/taubsie/dungeonhub/application/command/commands/CdnCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ContentConnection;
import me.taubsie.dungeonhub.application.enums.EmbedColor;
import me.taubsie.dungeonhub.application.enums.KnownStaticResource;
import me.taubsie.dungeonhub.application.exceptions.MissingPermissionException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.application.service.PermissionService;
import org.javacord.api.entity.Attachment;
import org.javacord.api.entity.message.component.ActionRowBuilder;
import org.javacord.api.entity.message.component.ButtonBuilder;
import org.javacord.api.entity.message.component.ButtonStyle;
import org.javacord.api.entity.message.embed.EmbedBuilder;
import org.javacord.api.entity.user.User;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandInteractionOption;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@CommandParameters(name = "cdn", description = "Manage the CDN", enabledForUsers = {356134481452597250L,
        531094512819240960L, 564353701003657216L, 574048571364605992L, 1116284449190064220L, 795048346955677748L,
        884589309037011015L, 346292488837005334L},
        enabledInDms = true)
public class CdnCommand extends Command {
    @Override
    public long[] getEnabledServers() {
        return new long[]{693263712626278553L, 1023684107877761196L};
    }

    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        User user = getUser();

        if (!PermissionService.getInstance().mayUseCdn(user)) {
            throw new MissingPermissionException();
        }

        SlashCommandInteractionOption firstOption =
                getOptionAtIndex(slashCommandCreateEvent.getSlashCommandInteraction(), 0);

        if (firstOption.getName().equalsIgnoreCase("static")) {
            KnownStaticResource resource = getEnumOption(firstOption, "file", KnownStaticResource.class);

            String url = ContentConnection.getInstance().getApiUrl(resource.getPath()).toString();

            respondEphemeral(ApplicationService.getInstance()
                            .getEmbed()
                            .setColor(EmbedColor.POSITIVE.getColor())
                            .setTitle("Static resource")
                            .addInlineField("Name", resource.getDisplayName())
                            .addInlineField("File name", resource.getName())
                            .addField("Full URL", url)
                            .setImage(url),
                    new ActionRowBuilder().addComponents(
                            new ButtonBuilder().setStyle(ButtonStyle.LINK)
                                    .setUrl(url)
                                    .setLabel("Open")
                                    .build()
                    ).build());

            return;
        }

        Attachment fileOption = getAttachmentOption(firstOption, "file");

        Optional<String> nameOption = getOptionalStringOption(firstOption, "name");

        slashCommandCreateEvent.getSlashCommandInteraction()
                .respondLater(true)
                .thenAccept(updater -> {
                    Optional<String> fileUrl;

                    if (nameOption.isPresent()) {
                        fileUrl = ContentConnection.getInstance()
                                .uploadFile(fileOption.asByteArray().join(), nameOption.get());
                    } else {
                        fileUrl = ContentConnection.getInstance()
                                .uploadFile(fileOption.asByteArray().join());
                    }

                    EmbedBuilder embedBuilder;
                    if (fileUrl.isPresent()) {
                        String url = ContentConnection.getInstance().getApiUrl(fileUrl.get()).toString();

                        embedBuilder = ApplicationService.getInstance()
                                .getEmbed()
                                .setTitle("File added.")
                                .setColor(EmbedColor.POSITIVE.getColor())
                                .setImage(url)
                                .setFooter(ApplicationService.getInstance()
                                        .getUnstableFooter())
                                .setTimestamp(null)
                                .addField("URL", url);

                        updater.addComponents(new ActionRowBuilder()
                                .addComponents(new ButtonBuilder()
                                        .setStyle(ButtonStyle.LINK)
                                        .setUrl(url)
                                        .setLabel("Click to open")
                                        .build())
                                .build());
                    } else {
                        embedBuilder = ApplicationService.getInstance()
                                .getErrorEmbed()
                                .setDescription("Couldn't upload media file.");
                    }

                    updater.addEmbed(embedBuilder).update();
                });
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption attachmentOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.ATTACHMENT)
                .setName("file")
                .setDescription("The file to add.")
                .setRequired(true)
                .build();

        SlashCommandOption nameOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("name")
                .setDescription("The name of the file.")
                .setRequired(false)
                .build();

        SlashCommandOption addOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("add")
                .setDescription("Add a file to the CDN.")
                .setOptions(List.of(attachmentOption, nameOption))
                .build();

        SlashCommandOptionBuilder fileOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.STRING)
                .setName("file")
                .setDescription("The static file to get.")
                .setRequired(true);

        Arrays.stream(KnownStaticResource.values()).forEach(ressource -> fileOption.addChoice(ressource.getDisplayName(), ressource.name()));

        SlashCommandOption staticOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.SUB_COMMAND)
                .setName("static")
                .setDescription("Show the list of static files of the CDN.")
                .setOptions(List.of(fileOption.build()))
                .build();

        return List.of(addOption, staticOption);
    }
}

File: src/main/java/me/taubsie/dungeonhub/kord/application/commands/CdnCommand.kt
package me.taubsie.dungeonhub.kord.application.commands

import com.kotlindiscord.kord.extensions.commands.Arguments
import com.kotlindiscord.kord.extensions.commands.application.slash.converters.impl.enumChoice
import com.kotlindiscord.kord.extensions.commands.application.slash.ephemeralSubCommand
import com.kotlindiscord.kord.extensions.commands.converters.impl.attachment
import com.kotlindiscord.kord.extensions.commands.converters.impl.optionalString
import com.kotlindiscord.kord.extensions.components.components
import com.kotlindiscord.kord.extensions.components.linkButton
import com.kotlindiscord.kord.extensions.extensions.Extension
import com.kotlindiscord.kord.extensions.extensions.ephemeralSlashCommand
import dev.kord.rest.builder.message.EmbedBuilder
import me.taubsie.dungeonhub.application.connection.DungeonHubConnection
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ContentConnection
import me.taubsie.dungeonhub.kord.application.enums.EmbedColor
import me.taubsie.dungeonhub.kord.application.enums.KnownStaticResource
import me.taubsie.dungeonhub.kord.application.exceptions.CommandExecutionException
import me.taubsie.dungeonhub.kord.application.loader.LoadExtension
import me.taubsie.dungeonhub.kord.application.service.ApplicationService
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.Request

@LoadExtension
class CdnCommand : Extension() {
    override val name = "cdn-command"
    private val allowedUsers = listOf(
        356134481452597250L,
        531094512819240960L,
        564353701003657216L,
        574048571364605992L,
        1116284449190064220L,
        795048346955677748L,
        884589309037011015L,
        346292488837005334L
    )

    override suspend fun setup() {
        ephemeralSlashCommand {
            name = "cdn"
            description = "Manage the CDN"
            allowInDms = true
            check {
                failIfNot("You aren't allowed to use this command.") {
                    allowedUsers.contains(event.interaction.user.id.value.toLong())
                }
            }

            ephemeralSubCommand(::AddArguments) {
                name = "add"
                description = "Add a file to the CDN."

                action {
                    respond {
                        val attachmentRequest = Request.Builder().url(arguments.attachment.url.toHttpUrl()).build()

                        val attachmentData: ByteArray =
                            DungeonHubConnection.getInstance().executeRawRequest(attachmentRequest)
                                .orElseThrow { CommandExecutionException("Couldn't read file data.") }

                        val fileUrl = if (arguments.name != null) {
                            ContentConnection.getInstance().uploadFile(attachmentData, arguments.name)
                        } else {
                            ContentConnection.getInstance().uploadFile(attachmentData)
                        }

                        val embedBuilder: EmbedBuilder
                        if (fileUrl != null && fileUrl.isPresent) {
                            val url = ContentConnection.getInstance().getCdnUrl(fileUrl.get()).toString()

                            embedBuilder = ApplicationService.embed
                            embedBuilder.title = "File added."
                            embedBuilder.color = EmbedColor.POSITIVE.color
                            embedBuilder.image = url
                            embedBuilder.footer {
                                text = ApplicationService.unstableFooter
                            }
                            embedBuilder.timestamp = null
                            embedBuilder.field("URL") { url }

                            components {
                                linkButton {
                                    label = "Click to open"
                                    this.url = url
                                }
                            }
                        } else {
                            embedBuilder = ApplicationService.errorEmbed
                            embedBuilder.description = "Couldn't upload media file."
                        }

                        embeds = mutableListOf(embedBuilder)
                    }
                }
            }

            ephemeralSubCommand(::StaticArguments) {
                name = "static"
                description = "Show the list of static files of the CDN."

                action {
                    respond {
                        val url = ContentConnection.getInstance().getStaticUrl(arguments.resource.path).toString()

                        val embed = ApplicationService.embed
                        embed.color = EmbedColor.POSITIVE.color
                        embed.title = "Static resource"
                        embed.field("Name", true) { arguments.resource.loadDisplayName() }
                        embed.field("File name", true) { arguments.resource.getName() }
                        embed.field("Full URL", false) { url }
                        embed.image = url

                        embeds = mutableListOf(embed)

                        components {
                            linkButton {
                                label = "Open"
                                this.url = url
                            }
                        }
                    }
                }
            }
        }
    }

    inner class AddArguments : Arguments() {
        val attachment by attachment {
            name = "file"
            description = "The file to add."
        }

        val name by optionalString {
            name = "name"
            description = "The name of the file."
        }
    }

    inner class StaticArguments : Arguments() {
        val resource by enumChoice<KnownStaticResource> {
            name = "file"
            description = "The static file to get."
            typeName = "KnownStaticResource"
        }
    }
}